[storybook] steps table component

dvdherron opened the pull request:
![](https://source.unsplash.com/featured/?cute,animal&thurs)
<!-- Swap `CHANGE_ME` for a random string to get a different image -->



## Description
Adds stepsTable component to the component library


## Steps to test/reproduce
- `yarn storybook`  or ` ./derrick storybook`
- see the component under the Plans section
- 
_Please explain how to best reproduce the issue and/or test the changes locally (including the pages/URLs/views/states to review)._


## Show me
_Provide screenshots/animated gifs/videos if necessary._


# REMEMBER: Attach this PR to the Trello card

dvdherron commented:
![](https://github.trello.services/images/mini-trello-icon.png) [MetaDeploy plans › stepsTable](https://trello.com/c/fVwIyByP/78-metadeploy-plans-stepstable)

dvdherron commented:
@jgerigmeyer I'm still working on the documentation for this component and it's partially rendered. 

Questions: 
  - What else is needed to get the logs to expand when you click the "show logs" button?
  - For showing the different states, there should probably be options for separately switching the job _and_ the steps, right?
  
I think previously you mentioned possibly needing to make a fake scratch org to get this to behave correctly?

jgerigmeyer commented:
> * What else is needed to get the logs to expand when you click the "show logs" button?

For some reason, the class-based child components weren't working right in Storybook. I converted those to functional components, and that mostly fixed it. The other issue was that the job was in "running" state but there was no active step -- and the logic generally assumes the job is either done or has at least one step without `results` or `status`. I switched to a "completed" job for now, tho I think ideally we want a dropdown to choose between different types. But the logs toggle seems to be working now.

> * For showing the different states, there should probably be options for separately switching the job _and_ the steps, right?

Hm, that's a good question. I guess we want to at least toggle between jobs in different states. But as far as the user interface for the story, I think one dropdown should do it, with choices e.g. "Complete", "Canceled", "Started", "Failed" etc. But you're right that the selection there will probably impact both the `job` and `steps` -- there's some redundancy/overlap there between plan/job/steps.

> I think previously you mentioned possibly needing to make a fake scratch org to get this to behave correctly?

No, that was in reference to the upcoming guided-tour story on Meteco.

dvdherron commented:
@jgerigmeyer @stacyk This is ready for review. 

@jgerigmeyer The logs are mostly how they should. One thing that isn't behaving correctly is when a step isn't required. The checkbox doesn't show up for some reason?

stacyk commented:
@dvdherron I added a few checklist items to the [Trello card](https://trello.com/c/fVwIyByP/78-metadeploy-plans-stepstable) from the meeting.

jgerigmeyer reviewed (commented):
@dvdherron This is looking good! I think there are only two things missing:

1. Let's use more realistic step names/types and result logs -- maybe copy some directly from production?
2. The prose description discusses how this component is used both for preflight / selecting steps, then also while the installation is running. But our example story only shows it in the context of an installation/job. That's probably fine for this PR, but I think we should add another story that demonstrates various states of the table as it's used on the Plan Detail page pre-installation?

dvdherron commented:
> 
> 
> @dvdherron This is looking good! I think there are only two things missing:
> 
>     1. Let's use more realistic step names/types and result logs -- maybe copy some directly from production?

        Cleaned these up with logs / names from production. 

       And it makes sense to add all of the steps for each status. I think step 3 was missing, so I adding that one in as well. 

> 
>     2. The prose description discusses how this component is used both for preflight / selecting steps, then also while the installation is running. But our example story only shows it in the context of an installation/job. That's probably fine for this PR, but I think we should add another story that demonstrates various states of the table as it's used on the Plan Detail page pre-installation?

Aha, so we would add a separate story for showing the table along with a pre-install validation button? Makes sense. I'll add a story.

dvdherron commented:
> 
> 
> > Cleaned these up with logs / names from production.
> 
> Looks good! 👍
> 
> > And it makes sense to add all of the steps for each status. I think step 3 was missing, so I adding that one in as well.
> 
> I had intentionally removed that, to show what a "skipped" optional step looks like. Not all steps are always chosen to be installed.

Aha, I'll put this back. 

> > Aha, so we would add a separate story for showing the table along with a pre-install validation button @jgerigmeyer ? Makes sense. I'll add a story.
> 
> Sort of -- I'm not sure we need to show the button in addition to the table though. I'm saying we should show the table in the pre-install state; right now we're only showing various installation states.
> 
> We're not showing the table in any state from the Plan Detail page, e.g. https://install.salesforce.org/products/npsp/latest/install
> 
> So that could happen in this PR (it seems like it's sort of part of the same story, to show the StepsTable in all the ways it's used), but could also be a new PR.

And work on add this state as well.

jgerigmeyer commented:
> Aha, I'll put this back.

I already did 😬 🏎️

stacyk commented on src/stories/plans/stepsTable.stories.mdx:
```diff
@@ -0,0 +1,44 @@
+When an installation **Step** is complete, a round, green icon with a white
+checkmark appears in the `Install` column.
+
+A toggle button for hiding or expanding all logs appears in the `Steps` column
+once the first **Step** starts to transmit log data.
```
Also, I see this show all logs toggle doesn't actually work on the example on this page. Is that an error or expected? I know the text says this button appears after the first step starts, but on the example page, it seems to only work if I change it to failed or complete.

jgerigmeyer replied:
> Also, I see this show all logs toggle doesn't actually work on the example on this page. Is that an error or expected? I know the text says this button appears after the first step starts, but on the example page, it seems to only work if I change it to failed or complete.

It was working "correctly" -- when a job is running, that toggle only expands the logs of the currently-running step, if it has any -- but our sample case didn't have any logs for the active step. Fixed now.

stacyk reviewed (commented):
looking good so far @dvdherron! I added a few suggestions for the copy. I'll check it out again next week if you add the other table states in this branch.

dvdherron commented:
@jgerigmeyer @stacyk 

Added examples of the pre-validation (or pre-install if no validation check is needed)
![image](https://user-images.githubusercontent.com/41588129/134187456-2c83ce3a-bace-4505-aca6-7949c510a7d4.png)

And when it's possible to start the installation process (Can Be Installed?)
![image](https://user-images.githubusercontent.com/41588129/134187657-33cb4132-1629-4660-bdf1-924fb96bef3e.png)

Changed the title of the default table to Active Job and switched the default job to 'Failed' since that one shows all of the possible statuses of a step (complete, skipped, failed). 

Unless I'm missing something, I don't think there is a table state for `failed` or `warn` validation checks, so this should cover everything? Let me know if that's wrong @jgerigmeyer

jgerigmeyer reviewed (approved):
@dvdherron Looking good! I tweaked a bit of text, and added more preflight result states.

> Changed the title of the default table to Active Job and switched the default job to 'Failed' since that one shows all of the possible statuses of a step (complete, skipped, failed).

I'm torn on this. I changed it back because there's something satisfying about the default example not having errors, and being (hopefully) the most common case. I also think showing an active step with some steps having logs and others not is nice.

> Unless I'm missing something, I don't think there is a table state for `failed` or `warn` validation checks, so this should cover everything? Let me know if that's wrong @jgerigmeyer

I added the remaining states: skipped, optional, warn, and error.

dvdherron commented:
> @dvdherron Looking good! I tweaked a bit of text, and added more preflight result states.
> 
> > Changed the title of the default table to Active Job and switched the default job to 'Failed' since that one shows all of the possible statuses of a step (complete, skipped, failed).
> 
> I'm torn on this. I changed it back because there's something satisfying about the default example not having errors, and being (hopefully) the most common case. I also think showing an active step with some steps having logs and others not is nice.

Makes sense. And it's easy to select the different states on the component example page. 

@jgerigmeyer The storybook controls for `plan` and `steps` are visible for `Needs pre-install validation` and `Pre-install validation complete`. Is that on purpose?

> > Unless I'm missing something, I don't think there is a table state for `failed` or `warn` validation checks, so this should cover everything? Let me know if that's wrong @jgerigmeyer
> 
> I added the remaining states: skipped, optional, warn, and error.

Oy. I missed these. I didn't realize there were also visible pre-flight results for each individual step as well. I thought if the pre-install validation failed nothing changed in the table. My bad.

dvdherron commented on src/js/components/plans/preflightResults.tsx:
```diff
@@ -41,13 +41,13 @@ export const JobError = ({ errors }: { errors: StepResult[] }) => {
   const warnList = [];
   let listItem = null;
-  for (const err of errors) {
+  for (const [idx, err] of errors.entries()) {
```
@jgerigmeyer does this change make it possible to see more than one error or warning message?

jgerigmeyer replied:
No, I don't think this has any functional impact. I just noticed a warning in the console about missing a [unique `key` on each item in the list](https://reactjs.org/docs/lists-and-keys.html).

jgerigmeyer commented:
> @jgerigmeyer The storybook controls for `plan` and `steps` are visible for `Needs pre-install validation` and `Pre-install validation complete`. Is that on purpose?

@dvdherron It was on purpose, but I could go either way there. I guess I thought it was more helpful to see the full object rather than just a `-` an didn't hurt anything, but I could be convinced otherwise if you think it's distracting.

stacyk commented:
> @dvdherron It was on purpose, but I could go either way there. I guess I thought it was more helpful to see the full object rather than just a `-` an didn't hurt anything, but I could be convinced otherwise if you think it's distracting.

I was also going to ask about this. I suppose it is nice to see the step names in case you want to change any ... but it was a little weird seeing the plan name since that isn't something that is seen (I know it's functional purpose).

stacyk commented on src/stories/plans/stepsTable.stories.mdx:
```diff
@@ -0,0 +1,48 @@
+if the installation is not running) appears in the `Steps` column header once
+the first **Step** starts to transmit log data.
+
+Skipped **Steps** display a gray dash in the `Install` column. If a **Step**
+fails its status is labeled as `Error` and an error message is shown below the
+**Step** name.
+
```
@dvdherron maybe since we don't show this (I think you mentioned you did but @jgerigmeyer thought it would be best to show the non-error states on the landing page) we could add a relative link to the error example perhaps?
 `/path=/story/plans-stepstable-examples--post-validation&args=preflight:Error`

dvdherron replied:
> 
> 
> @dvdherron maybe since we don't show this (I think you mentioned you did but @jgerigmeyer thought it would be best to show the non-error states on the landing page) we could add a relative link to the error example perhaps?
> `/path=/story/plans-stepstable-examples--post-validation&args=preflight:Error`


I tried this @stacyk , but since that addition at the end (after `--post-validation`) doesn't match any of the id's the component doesn't render correctly. I think we would have to make a separate "Failed" example to render it inline like this.

stacyk replied:
@dvdherron Ah, yes, I see. I think it is fine as is. If it is unclear they can jump to the component and play with the controls :)

stacyk reviewed (approved):
Looks great

stacyk commented:
> > @dvdherron It was on purpose, but I could go either way there. I guess I thought it was more helpful to see the full object rather than just a `-` an didn't hurt anything, but I could be convinced otherwise if you think it's distracting.
> 
> I was also going to ask about this. I suppose it is nice to see the step names in case you want to change any ... but it was a little weird seeing the plan name since that isn't something that is seen (I know it's functional purpose).

When I change the status to error on the Pre Install Validation complete and then go to docs, the table shows that error + warning state on the docs page. I think that surprised me but it isn't something bad. Is that expected?

@dvdherron because we talk about the status symbols for failed, complete, skipped, etc do you think we should mention the warning state?

jgerigmeyer commented:
> When I change the status to error on the Pre Install Validation complete and then go to docs, the table shows that error + warning state on the docs page. I think that surprised me but it isn't something bad. Is that expected?

Yes, the docs page shows the latest state of the story, so it updates if you change the settings on the story itself.

dvdherron commented:
> 
> 
> > > @dvdherron It was on purpose, but I could go either way there. I guess I thought it was more helpful to see the full object rather than just a `-` an didn't hurt anything, but I could be convinced otherwise if you think it's distracting.
> > 
> > 
> > I was also going to ask about this. I suppose it is nice to see the step names in case you want to change any ... but it was a little weird seeing the plan name since that isn't something that is seen (I know it's functional purpose).
> 
> When I change the status to error on the Pre Install Validation complete and then go to docs, the table shows that error + warning state on the docs page. I think that surprised me but it isn't something bad. Is that expected?
> 
> @dvdherron because we talk about the status symbols for failed, complete, skipped, etc do you think we should mention the warning state?

Adding in a line for a warning after a pre-install validation @stacyk. For a Job 

> 
> 
> > > @dvdherron It was on purpose, but I could go either way there. I guess I thought it was more helpful to see the full object rather than just a `-` an didn't hurt anything, but I could be convinced otherwise if you think it's distracting.
> > 
> > 
> > I was also going to ask about this. I suppose it is nice to see the step names in case you want to change any ... but it was a little weird seeing the plan name since that isn't something that is seen (I know it's functional purpose).
> 
> When I change the status to error on the Pre Install Validation complete and then go to docs, the table shows that error + warning state on the docs page. I think that surprised me but it isn't something bad. Is that expected?
> 
> @dvdherron because we talk about the status symbols for failed, complete, skipped, etc do you think we should mention the warning state?

@stacyk I added a mention of the warning and error states for `Step`s post install-validations.

For Installation Jobs, is it accurate that a `Step` can have a `Warning` _after_ an installation has finished @jgerigmeyer ?

jgerigmeyer commented:
> For Installation Jobs, is it accurate that a `Step` can have a `Warning` _after_ an installation has finished @jgerigmeyer ?

@dvdherron No. Job results can only be `ok` or `error`. Preflight results can be `error`, `warn`, `skip`, `optional` or `hide`.

stacyk commented:
> @stacyk I added a mention of the warning and error states for `Step`s post install-validations.

Looks good.

jgerigmeyer commented:
@dvdherron @stacyk I dug into this a bit more, and pushed a change that displays more states in the docs page. Better or worse?

stacyk commented:
> @dvdherron @stacyk I dug into this a bit more, and pushed a change that displays more states in the docs page. Better or worse?

I'm a fan of visual examples so I like that we are including the additional states on the docs page.

dvdherron commented:
> 
> 
> @dvdherron @stacyk I dug into this a bit more, and pushed a change that displays more states in the docs page. Better or worse?

Looks good to me. 

Ah, and the `table` option you've disabled is preventing those properties from showing up at all in the controls table? good to know that's an option.

jgerigmeyer commented on src/stories/plans/stepsTable.stories.mdx:
```diff
@@ -16,41 +16,58 @@ used to display information about each **Step** involved in installing a
+it is optional, a selectable checkbox is shown instead.
+
+<Story
+  id="plans-stepstable-examples--plan-steps&args=preflight:Success"
+  height="320px"
```
@dvdherron For some reason, when we embed a story with `args` like this, it renders in an `iframe` -- which is okay, but means we need to manually set the `height`.

jgerigmeyer commented on src/stories/plans/stepsTable.stories.tsx:
```diff
@@ -27,71 +27,56 @@ const sampleJobs: { [key: string]: Job } = {
 };
-const samplePreflights: { [key: string]: Preflight } = {
+const samplePreflights: { [key: string]: Preflight | null } = {
+  None: null,
```
I went ahead and combined the various Preflight examples into one story, so now we have one for pre-install and one for installation.

jgerigmeyer commented:
@dvdherron If this looks good to you, I'd say go ahead and merge it 🚀

dvdherron commented:
> 
> 
> @dvdherron If this looks good to you, I'd say go ahead and merge it 🚀

🎆 Ok, fixed a tiny typo. merging after these tests pass.
